File: packages/plugin-contract/src/capability.ts
import type { ActionSeverity } from "./severity";
import type { JsonSchema } from "./jsonSchema";

export interface CapabilityDescriptor {
  name: string;
  description: string;
  severity: ActionSeverity;
  supportsDryRun: boolean;
  /**
   * JSON Schema real (docs/16 P1). Validado antes de invocar el driver,
   * en dos capas: `ToolResolver` (Gateway) y `CapabilityRegistry` (Edge
   * Agent) — ver `validateAgainstSchema` en `./schemaValidation`. Cubre
   * solo forma y tipos básicos; las reglas de negocio (rangos, formatos,
   * valores permitidos) siguen viviendo en el propio plugin.
   */
  inputSchema?: JsonSchema;
  /**
   * Nombre del campo del input que identifica el target físico afectado
   * (ej. "pin" para un GPIO). Habilita que la Safety Policy del dispositivo
   * (SafetyPolicyStore, @kan/edge-agent-core) sobrescriba la severidad de
   * esta capability por target individual. Sin este campo, la capability
   * nunca es afectada por overrides — usa siempre `severity` tal cual.
   */
  targetParam?: string;
}

export interface CapabilityResult {
  success: boolean;
  data?: unknown;
  error?: string;
}

File: packages/kan-plugin-sdk-py/src/kan_plugin_sdk_py/capability.py
"""Espejo a mano de packages/plugin-contract/src/capability.ts.

`input_schema` viaja como dict JSON Schema tal cual — la validación real
sigue ocurriendo del lado TypeScript (`CapabilityRegistry` en el Edge
Agent, vía `validateAgainstSchema`) antes de que `invoke` llegue al
sidecar por el protocolo WS; este SDK no reimplementa esa validación.
"""

from dataclasses import dataclass
from typing import Any, Optional

from .severity import ActionSeverity


@dataclass
class CapabilityDescriptor:
    name: str
    description: str
    severity: ActionSeverity
    supports_dry_run: bool
    input_schema: Optional[dict[str, Any]] = None
    target_param: Optional[str] = None

    def to_wire(self) -> dict[str, Any]:
        """Forma exacta que espera `CapabilityDescriptor` del lado TS (camelCase)."""
        wire: dict[str, Any] = {
            "name": self.name,
            "description": self.description,
            "severity": self.severity,
            "supportsDryRun": self.supports_dry_run,
        }
        if self.input_schema is not None:
            wire["inputSchema"] = self.input_schema
        if self.target_param is not None:
            wire["targetParam"] = self.target_param
        return wire


@dataclass
class CapabilityResult:
    success: bool
    data: Optional[Any] = None
    error: Optional[str] = None

    def to_wire(self) -> dict[str, Any]:
        wire: dict[str, Any] = {"success": self.success}
        if self.data is not None:
            wire["data"] = self.data
        if self.error is not None:
            wire["error"] = self.error
        return wire
